### PR TITLE
Encode repeating

### PR DESCRIPTION
**Pull request by shrutivanw**
CS Fundamentals' assignment to replace repeating characters in the input string with character followed by character count.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,12 @@
-# encode_repeating
\ No newline at end of file
+# Encode Repeating
```
Comment (Hamled) [inline README.md:1]: Should we link them to the [Wikipedia article for run-length encoding](https://en.wikipedia.org/wiki/Run-length_encoding) for more context around this idea?

File: README.md [hunk 2]
```diff
@@ -1 +1,12 @@
...
+
+## Exercise
+* Design and implement a method that updates the string by replacing consecutive repeating characters with a number representing the frequncy. The replacement is done only if the string length will get reduced by the process. In addition to `.length`, you may use `.slice!` to implement this method.
+   - e.g "aaabbbbbcccc" will get replace to "a3b5c4", since a is repeated 3 times, b is repeated 5 times and c is repeated 4 times
+   - "xxxyttttgeee" will get replaced with "x3yt4ge3". 'y' and 'g' do not get replaced by "y1" and "g1" since that will increase the length.
+   - "ddbbfffgjjjj" will get replaced with "ddbbf3gj4". "dd" and "bb" do not get replaced since the length will remain the same with "d1" and "b1". "g" does not get replaced with "g1" since that will increase the length.
+* Share and explain the time and space complexities for your solution.
+* At minimum, your implementation should pass the basic tests.
+**Note**: Do not use Ruby provided functionality for `.reverse` and `.reverse!`. You may use `.length` and `.slice!` to implement this method.
```
Comment (Hamled) [inline README.md]: We should have an extra blank line above this so that the note goes into its own paragraph.

File: lib/encode_repeating.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# A method that updates the string by replacing consecutive repeating characters
+# with a number representing the frequency. The replacement is done only if the
+# string length will get reduced by this process.
+def encode_repeating(my_string)
+  my_string << "NOT IMPLEMENTED"
```
Comment (Hamled) [inline lib/encode_repeating.rb]: How would you feel about doing `raise NotImplementedError` here instead?

Best practice I think is generally to avoid `puts` statements and other things that will print to the console for code being tested, because that output kind of gets randomly interspersed with the test runner's output.

File: specs/encode_repeating_spec.rb [hunk 4]
```diff
@@ -0,0 +1,82 @@
+require 'minitest/autorun'
+require 'minitest/reporters'
+require_relative '../lib/encode_repeating'
+
+describe "encode repeating" do
+  describe "basic tests" do
+    describe "test 1" do
```
Comment (Hamled) [inline specs/encode_repeating_spec.rb]: Same as the other repos, we should probably remove the `describe` blocks that are immediate parents of individual tests.

File: specs/encode_repeating_spec.rb [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        encode_repeating(test_string)
+
+        test_string.must_equal "ddbbf3gj4"
+      end
+    end
+  end
+
+  # if it's a string parameter, check for empty
+  describe "empty string input test" do
+    it "empty string remains as is" do
```
Comment (Hamled) [inline specs/encode_repeating_spec.rb]: I think we could move this test and the nil test into the edge cases section.

File: specs/encode_repeating_spec.rb [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+      test_string.must_be_nil
+    end
+  end
+
+  # check for edge cases
+  describe "edge cases" do
+    describe "edge case 1" do
+      it "string with no consecutive repeating characters" do
+        test_string = "at53fag"
```
Comment (Hamled) [inline specs/encode_repeating_spec.rb]: I think this is a good edge case test, however I have two concerns:
1. Do we have specific instructions for when the input string includes digits? I think the simplest thing would be to have no tests with digits, or to expect that `ArgumentError` is raised when the input string has digits.
1. I'm sure it was unintentional, but we should remove the slur from this test string.

**Review by Hamled — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (shrutivanw) [reply README.md:1]: That's a great link! Let me add it as a further reading link.

File: README.md (hunk 2 above)
Comment (shrutivanw) [reply README.md]: Done!

File: lib/encode_repeating.rb (hunk 3 above)
Comment (shrutivanw) [reply lib/encode_repeating.rb]: Great idea!

File: specs/encode_repeating_spec.rb (hunk 6 above)
Comment (shrutivanw) [reply specs/encode_repeating_spec.rb]: Whoops! Good catch on the slur - that didn't even occur to me as typed random letters. Oh boy! Thank you!
I changed the test string for the no consecutive repeating character to be "aeiou"

**Comment by shrutivanw** [comment]
Thank you @Hamled PR updated.

**Review by Hamled — APPROVED** [review]
